File: utils/pii.py
"""PII masking for analyst surfaces (API responses and UI display).

Only Admin may see full customer PII. All other roles receive masked values.
"""
from typing import Any, Dict, Mapping, Optional

# Map record keys → display_pii field names
PII_FIELD_MAP: Mapping[str, str] = {
    "email": "email",
    "phone_number": "phone",
    "phone": "phone",
    "address": "address",
    "street": "street",
    "ip_address": "ip",
    "ip": "ip",
}


def can_view_full_pii(analyst: Optional[Dict[str, Any]]) -> bool:
    """Only Admin may see full customer PII in the analyst portal."""
    role = (analyst or {}).get("role", "")
    return role == "Admin"


def mask_email(email: str) -> str:
    """Mask local part after the first two characters: er*******@gmail.com."""
    if not email or "@" not in email:
        return email or ""
    local, domain = email.split("@", 1)
    if not local:
        return f"***@{domain}"
    if len(local) == 1:
        masked_local = "*"
    elif len(local) == 2:
        masked_local = local[0] + "*"
    else:
        masked_local = local[:2] + "*" * (len(local) - 2)
    return f"{masked_local}@{domain}"


def mask_phone(phone: str) -> str:
    """Keep first and last two characters: 91******52."""
    if not phone:
        return ""
    digits = str(phone).strip()
    if len(digits) <= 4:
        return "***"
    return digits[:2] + "*" * (len(digits) - 4) + digits[-2:]


def mask_street(street: str) -> str:
    """Mask street like email local part: keep first two chars, star the rest."""
    if not street:
        return ""
    value = street.strip()
    if len(value) <= 2:
        return "*" * len(value)
    return value[:2] + "*" * (len(value) - 2)


def mask_address(address: str) -> str:
    """Mask the street portion of an address; leave city/state/zip visible."""
    if not address:
        return ""
    value = address.strip()
    if "," in value:
        street, rest = value.split(",", 1)
        return f"{mask_street(street)},{rest}"
    return mask_street(value)


def mask_ip(ip_address: str) -> str:
    if not ip_address:
        return ""
    parts = ip_address.split(".")
    if len(parts) == 4:
        return f"{parts[0]}.{parts[1]}.***.***"
    return "***"


def display_pii(
    value: str,
    *,
    field: str,
    analyst: Optional[Dict[str, Any]] = None,
) -> str:
    """Show full PII for Admin; mask for all other analyst roles."""
    if not value:
        return ""
    if can_view_full_pii(analyst):
        return str(value)
    maskers = {
        "email": mask_email,
        "phone": mask_phone,
        "address": mask_address,
        "street": mask_street,
        "ip": mask_ip,
    }
    mask_fn = maskers.get(field)
    return mask_fn(value) if mask_fn else value


def sanitize_pii_record(
    record: Optional[Dict[str, Any]],
    analyst: Optional[Dict[str, Any]] = None,
) -> Optional[Dict[str, Any]]:
    """Return a copy of *record* with PII fields masked unless analyst is Admin.

    Use at the API / fetch boundary so non-admin sessions never hold raw PII.
    """
    if record is None:
        return None
    out = dict(record)
    if can_view_full_pii(analyst):
        return out
    for key, field in PII_FIELD_MAP.items():
        if key in out and out[key] is not None and out[key] != "":
            out[key] = display_pii(str(out[key]), field=field, analyst=analyst)
    return out

File: analyst-portal/src/pii.ts
/** Role-based PII display — mirrors utils/pii.py (Admin = full; others masked). */

export function canViewFullPii(analyst?: { role?: string } | null): boolean {
  return analyst?.role === "Admin";
}

export function maskEmail(email: string): string {
  if (!email || !email.includes("@")) return email || "";
  const [local, domain] = email.split("@");
  if (!local) return `***@${domain}`;
  let masked: string;
  if (local.length === 1) masked = "*";
  else if (local.length === 2) masked = `${local[0]}*`;
  else masked = local.slice(0, 2) + "*".repeat(local.length - 2);
  return `${masked}@${domain}`;
}

export function maskPhone(phone: string): string {
  if (!phone) return "";
  const digits = String(phone).trim();
  if (digits.length <= 4) return "***";
  return digits.slice(0, 2) + "*".repeat(digits.length - 4) + digits.slice(-2);
}

export function maskStreet(street: string): string {
  if (!street) return "";
  const value = street.trim();
  if (value.length <= 2) return "*".repeat(value.length);
  return value.slice(0, 2) + "*".repeat(value.length - 2);
}

export function maskAddress(address: string): string {
  if (!address) return "";
  const value = address.trim();
  const idx = value.indexOf(",");
  if (idx >= 0) return maskStreet(value.slice(0, idx)) + value.slice(idx);
  return maskStreet(value);
}

export function maskIp(ip: string): string {
  if (!ip) return "";
  const parts = ip.split(".");
  if (parts.length === 4) return `${parts[0]}.${parts[1]}.***.***`;
  return "***";
}

export function displayPii(
  value: unknown,
  field: "email" | "phone" | "address" | "ip",
  analyst?: { role?: string } | null,
): string {
  if (value == null || value === "") return "";
  const str = String(value);
  if (canViewFullPii(analyst)) return str;
  if (field === "email") return maskEmail(str);
  if (field === "phone") return maskPhone(str);
  if (field === "address") return maskAddress(str);
  return maskIp(str);
}
